Raw nnet3 training

vimalmanohar opened the pull request:
Scripts for raw nnet3 training supporting quadratic objective function and dense targets.
Let me know if this looks ok. 
@vijayaditya @danpovey

vijayaditya commented on egs/wsj/s5/steps/nnet3/components.py:
```diff
@@ -152,7 +152,16 @@ def AddSoftmaxLayer(config_lines, name, input):
 
 
-def AddOutputLayer(config_lines, input, label_delay=None, suffix=None):
+def AddSigmoidLayer(config_lines, name, input):
```
Add self_repair_scale option.

vijayaditya commented on egs/wsj/s5/steps/nnet3/components.py:
```diff
@@ -160,9 +169,9 @@ def AddOutputLayer(config_lines, input, label_delay=None, suffix=None):
+        component_nodes.append('output-node name={0} input={1} objective={2}'.format(name, input['descriptor'], objective_type))
     else:
-        component_nodes.append('output-node name={0} input=Offset({1},{2})'.format(name, input['descriptor'], label_delay))
+        component_nodes.append('output-node name={0} input=Offset({1},{2}) objective={3}'.format(name, input['descriptor'], label_delay, objective_type))
```
use named arguments in format()

vimalmanohar replied:
The rest of components.py does not have named arguments. Should I modify all of it?

vijayaditya replied:
If you have the time yes.

On Wed, Mar 2, 2016 at 2:38 AM, Vimal Manohar <notifications@github.com>
wrote:

> In egs/wsj/s5/steps/nnet3/components.py
> <https://github.com/kaldi-asr/kaldi/pull/539#discussion_r54688606>:
>
> >      else:
> > -        component_nodes.append('output-node name={0} input=Offset({1},{2})'.format(name, input['descriptor'], label_delay))
> > +        component_nodes.append('output-node name={0} input=Offset({1},{2}) objective={3}'.format(name, input['descriptor'], label_delay, objective_type))
>
> The rest of components.py does not have named arguments. Should I modify
> all of it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/539/files#r54688606>.
>

vijayaditya commented on egs/wsj/s5/steps/nnet3/get_egs.sh:
```diff
@@ -265,23 +293,70 @@ valid_egs_opts="--left-context=$valid_left_context --right-context=$valid_right_
-  rm $dir/.error 2>/dev/null
-  echo "$0: ... extracting validation and training-subset alignments."
 
+if ! $raw_nnet; then
```
the steps/nnet3/get_egs.sh is really complicated with the new changes. I would recommend using separate scripts.

vimalmanohar replied:
Agreed. What about steps/nnet3/tdnn/train.sh? Should I create a script like steps/nnet3/tdnn/train_raw.sh?

vijayaditya replied:
I was suggesting splitting the get_egs.sh script. Not the train script. If
you have time to modify the train script I would recommend adding the
changes to steps/nnet3/train_dnn.py.

On Wed, Mar 2, 2016 at 2:40 AM, Vimal Manohar <notifications@github.com>
wrote:

> In egs/wsj/s5/steps/nnet3/get_egs.sh
> <https://github.com/kaldi-asr/kaldi/pull/539#discussion_r54688722>:
>
> >
> > +if ! $raw_nnet; then
>
> Agreed. What about steps/nnet3/tdnn/train.sh? Should I create a script
> like steps/nnet3/tdnn/train_raw.sh?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/539/files#r54688722>.
>

vijayaditya replied:
I reason I would be fine with adding the changes to the same python script is due to the extensive use of functions in python script. It is usually easier to read. If you do feel the python script is also getting complicated due to excessive branching you could consider splitting the train.py script. Its up to you how you want to do this.

vimalmanohar replied:
Sure. Will look into this.

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/make_configs.py:
```diff
@@ -43,8 +43,16 @@ def GetArgs():
+                        help="skip lda layer", default=False, choices = ["false", "true"])
     parser.add_argument("--include-log-softmax", type=str, action=nnet3_train_lib.StrToBoolAction,
                         help="add the final softmax layer ", default=True, choices = ["false", "true"])
+    parser.add_argument("--add-final-sigmoid", type=str, action=nnet3_train_lib.StrToBoolAction,
```
group add-final-sigmoid and objective type options. Checkout how to groups arguments in argparse.

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/make_configs.py:
```diff
@@ -121,6 +129,10 @@ def CheckArgs(args):
         args.nonlin_input_dim = args.pnorm_input_dim
         args.nonlin_output_dim = args.pnorm_output_dim
 
+    if add_final_sigmoid and include_log_softmax:
```
there is something called mutually_exclusive_group in argparse. Use that rather than manually performing the check.

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/make_configs.py:
```diff
@@ -339,17 +355,25 @@ def MakeConfigs(config_dir, splice_indexes_string,
-                           prior_scale_file = prior_scale_file,
-                           include_log_softmax = include_log_softmax)
+
+        if not add_final_sigmoid:
```
it feels strange to check for the not of a boolean. Could you switch the check to 
if add_final_sigmoid:

Also add a brief local comment about why this is done (i.e., for quadratic objectives)

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/train.sh:
```diff
@@ -5,7 +5,7 @@
 #           2013  Xiaohui Zhang
 #           2013  Guoguo Chen
-#           2014  Vimal Manohar
+#           2014-2016  Vimal Manohar
```
it would be great if you could move this code to steps/nnet3/train_dnn.py. This will help us use the same script for a variety of architectures.

danpovey replied:
Vimal, I don't want to check in this code as-is.
You have made a huge number of changes to nnet3/tdnn/train.sh.  This creates problems for testing.
Obviously we're not going to re-run the million TDNN training setups that we have here and there in Kaldi.
And even apart from that, your changes obscure what was already a very complicated script.
I would possibly be OK with checking in a separate script whose purpose is to train DNNs for this type of setup.
Or you may be able to do it with Vijay's train_dnn.py, although again, I don't want a situation where you introduce
a ton of if statements into a program-- if there are a lot of changes, I'd prefer you to create a separate python version of the training script; this will mean we don't have to worry about introducing bugs into Vijay's working setup.
I don't have strong opinions about whether you check in a shell script or a python script, but I think it should be a separate script.  Right now your changes add to much complexity to the script.

danpovey replied:
BTW, you might find it easier to have the top-level train.sh script not be responsible for generating the configs or the egs-- if you assume that the top-level run.sh calls the appropriate scripts beforehand, that would make your life easier.  I'm following this pattern in the xvector setup.

vimalmanohar replied:
I can add the changes to a new script called nnet3/tdnn/train_raw.sh.

danpovey replied:
that would be OK with me, but try to simplify it by cutting out unnecessary
code.

On Thu, Mar 3, 2016 at 12:25 AM, Vimal Manohar <notifications@github.com>
wrote:

> In egs/wsj/s5/steps/nnet3/tdnn/train.sh
> <https://github.com/kaldi-asr/kaldi/pull/539#discussion_r54837404>:
>
> > @@ -5,7 +5,7 @@
> >  # Copyright 2012-2015  Johns Hopkins University (Author: Daniel Povey).
> >  #           2013  Xiaohui Zhang
> >  #           2013  Guoguo Chen
> > -#           2014  Vimal Manohar
> > +#           2014-2016  Vimal Manohar
>
> I can add the changes to a new script called nnet3/tdnn/train_raw.sh.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/539/files#r54837404>.
>

vimalmanohar commented:
I made most of the fixes suggested by @vijayaditya and fixed some bugs. It should be good for commit.
I reverted to not using the mutually exclusive group because the options are not really mutually exclusive (they only cannot both be true). 
I have left the suggestions to add this to train_dnn.py and to use named arguments in components.py to a future PR because they still require more time.

danpovey commented on egs/wsj/s5/steps/nnet3/get_egs_raw_nnet.sh:
```diff
@@ -0,0 +1,411 @@
+#!/bin/bash
```
I don't like the name of this program, because there is no inherent connection with a 'raw' neural net- the get_egs program doesn't need to see the neural net.  maybe call it get_egs_targets.sh, as in this case you supply more general targets, not alignments.

vimalmanohar commented:
I added a new script that deals only with raw nnet3 model and reads targets directly instead of ali-dir.

vijayaditya commented on egs/wsj/s5/steps/nnet3/components.py:
```diff
@@ -186,13 +196,34 @@ def AddFinalLayer(config_lines, input, output_dim,
-    AddOutputLayer(config_lines, prev_layer_output, label_delay, suffix = name_affix)
+    AddOutputLayer(config_lines, prev_layer_output, label_delay, suffix = name_affix, objective_type = objective_type)
+
+def AddFinalSigmoidLayer(config_lines, input, output_dim,
```
You should add a comment on what a FinalSigmoidLayer is and where it is intended to be used. I think the default objective_type for this function should be quadratic, as it will used primarily in that context.

vijayaditya commented on egs/wsj/s5/steps/nnet3/get_egs_targets.sh:
```diff
@@ -0,0 +1,411 @@
+if [ $stage -le 3 ]; then
+  echo "$0: Getting validation and training subset examples."
+  rm -f $dir/.error 2>/dev/null
+  echo "$0: ... extracting validation and training-subset alignments."
```
update the message

vijayaditya commented on egs/wsj/s5/steps/nnet3/get_egs_targets.sh:
```diff
@@ -0,0 +1,411 @@
+    # there are some extra soft links that we should delete.
+    for f in $dir/egs.*.*.ark; do rm $f; done
+  fi
+  echo "$0: removing temporary alignments and transforms"
```
update the message

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/make_configs.py:
```diff
@@ -121,6 +131,9 @@ def CheckArgs(args):
         args.nonlin_output_dim = args.pnorm_output_dim
 
+    if args.add_final_sigmoid and args.include_log_softmax:
+        sys.exit("--include-log-softmax and --add-final-sigmoid cannot both be true.")
```
Don't use <code>sys.exit</code> raise Exceptions.

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw.sh:
```diff
@@ -0,0 +1,636 @@
+cmvn_opts=  # will be passed to get_lda.sh and get_egs.sh, if supplied.
+            # only relevant for "raw" features, not lda.
+feat_type=raw  # or set to 'lda' to use LDA features.
+align_cmd=              # The cmd that is passed to steps/nnet2/align.sh
```
are you going to use alignments ? If not this realign code should not be here. trim it.

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw.sh:
```diff
@@ -0,0 +1,636 @@
+if [ -f path.sh ]; then . ./path.sh; fi
+. parse_options.sh || exit 1;
+
+if [ $# != 3 ]; then
```
update usage message to reflect your current options.

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw.sh:
```diff
@@ -0,0 +1,636 @@
+
+  config_opts=()
+
+  if [ ! -z "$relu_dim" ]; then
```
As @danpovey pointed out it would make your life a lot easier if you remove the config generation stage from this script. I place it in the top level run.sh script like [this](https://github.com/kaldi-asr/kaldi/blob/master/egs/swbd/s5c/local/nnet3/run_tdnn.sh#L52).

vimalmanohar replied:
I see. It is possible to move it outside and comment out the config generation to show an example.

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw.sh:
```diff
@@ -0,0 +1,636 @@
+
+# First work out the feature and iVector dimension, needed for tdnn config creation.
+case $feat_type in
+  raw) feat_dim=$(feat-to-dim --print-args=false scp:$data/feats.scp -) || \
```
while you are at it you could drop support for lda features from this script to make it cleaner. We are not using this option in nnet3 scripts as we use a preconditioning matrix in the network.

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw.sh:
```diff
@@ -0,0 +1,636 @@
+fi
+
+
+if $include_log_softmax && ! $dense_targets && [ $stage -le -2 ]; then
```
Is your code going into this branch given that you would not be using it with sparse targets any more ?

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw.sh:
```diff
@@ -0,0 +1,636 @@
+    "ark:nnet3-merge-egs ark:$cur_egs_dir/train_diagnostic.egs ark:- |" &
+fi
+
+if $include_log_softmax && [ $stage -le $[$num_iters+1] ]; then
```
Is this branch still necessary ?

vijayaditya commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw.sh:
```diff
@@ -0,0 +1,636 @@
+#!/bin/bash
```
I think the script should be named as <code>train_rawnnet_densetargets.sh</code>. Since there is another effort on raw features things can get pretty confusing in the future.

vimalmanohar replied:
I think the raw features must be called raw waveform features or something. "raw" here is for raw nnet without transition model (HMM) and that semantic is already being used in all the nnet3 binaries with --raw option. This script also supports sparse targets (e.g. SAD, speech / silence / music / noise classification or any other classification task that does not involve a HMM). The include_log_softmax is there for the same reason.

vijayaditya commented:
remember to edit the commit message to be more descriptive.

lifeiteng commented:
I don't think ` egs/wsj/s5/steps/nnet3/get_egs_targets.sh` have done all the job, if you use Sigmoid at last layer, you should also store the label info(dense target may be 1 or 0); 
when target value is 0, you will don't know to update which label's parameters(if target is for label 10, so the 0-9 Row parameters of last AffineTransform should not be updated), so you should also store the label info, maybe implement anothor method
```
  NnetIo(const std::string &name,
         int32 t_begin, const MatrixBase<BaseFloat> &feats, const Posterior &labels);
```

lifeiteng commented:
Also some changes in 
```
void NnetTrainer::ProcessOutputs(const NnetExample &eg,
                                 NnetComputer *computer) {
  std::vector<NnetIo>::const_iterator iter = eg.io.begin(),
      end = eg.io.end();
  for (; iter != end; ++iter) {
    const NnetIo &io = *iter;
    int32 node_index = nnet_->GetNodeIndex(io.name);
    KALDI_ASSERT(node_index >= 0);
    if (nnet_->IsOutputNode(node_index)) {
      ObjectiveType obj_type = nnet_->GetNode(node_index).u.objective_type;
      BaseFloat tot_weight, tot_objf;
      bool supply_deriv = true;
      // TODO use labels in NnetIo to Rewrite cross-entropy objective's derive (for sigmoid) in ComputeObjectiveFunction() 
      ComputeObjectiveFunction(io.features, obj_type, io.name,
                               supply_deriv, computer,
                               &tot_weight, &tot_objf);
      // TODO the accuracy's computation also should be changed.
      objf_info_[io.name].UpdateStats(io.name, config_.print_interval,
                                      num_minibatches_processed_++,
                                      tot_weight, tot_objf);
    }
  }
}
```

vimalmanohar commented:
@lifeiteng I am not sure what scenario you have in mind. This is the use case I have in mind:
The sigmoid is to be used as last layer if you want the output in each dimension to be like probabilities between zero and one. It is left to the user to decide the appropriate targets, which should be a vector with each element between 0 and 1. The objective function itself is independent of what is used in the network. Usually you can use quadratic for this but linear (which will be a cosine distance) might also ok.

vimalmanohar commented:
I have made changes based on the suggestions. I created an example run script in swbd/s5c

vimalmanohar commented on egs/wsj/s5/steps/nnet3/tdnn/make_configs.py:
```diff
@@ -357,11 +385,23 @@ def MakeConfigs(config_dir, splice_indexes_string,
     print('model_left_context=' + str(left_context), file=f)
     print('model_right_context=' + str(right_context), file=f)
     print('num_hidden_layers=' + str(num_hidden_layers), file=f)
+    print('num_targets=' + str(num_targets), file=f)
```
@vijayaditya 
I had to add extra options the vars file as they are required by the train script. It seems that the python library to parse the config file currently only supports integer options. Does it make sense to modify it to support generic options (string etc.)?

vijayaditya replied:
which python library are you talking about ?

vimalmanohar replied:
The function ParseModelConfigVarsFile in [this](https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/steps/nnet3/nnet3_train_lib.py). I want to modify this to return a python dict, which maps variables to strings by default, but integers for specific variables like model_left_context. This might require modifying all the python nnet3 train scripts.

vijayaditya replied:
I would recommend writing a new function tailored to your needs.

danpovey commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw_nnet.sh:
```diff
@@ -0,0 +1,626 @@
+num_jobs_compute_prior=10 # these are single-threaded, run on CPU.
+get_egs_stage=0    # can be used for rerunning after partial
+online_ivector_dir=
+presoftmax_prior_scale_power=-0.25
```
I think you can delete anything related to presoftmax_prior_scale.

danpovey commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw_nnet.sh:
```diff
@@ -0,0 +1,626 @@
+
+
+###
+# The config creation is moved to upper-level run script.
```
I don't see the need for this extended comment here.

danpovey commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw_nnet.sh:
```diff
@@ -0,0 +1,626 @@
+[ $num_jobs_final -gt $num_archives_expanded ] && \
+  echo "$0: --final-num-jobs cannot exceed #archives $num_archives_expanded." && exit 1;
+
+
```
you should probably check that add_lda is defined at this point and die if not.  otherwise we'll get confusing errors if something went wrong in generating the configs, or a different script was used.

danpovey commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw_nnet.sh:
```diff
@@ -0,0 +1,626 @@
+if $include_log_softmax && ! $dense_targets && [ $stage -le -2 ]; then
+  echo "$0: preparing initial vector for FixedScaleComponent before softmax"
+  echo "  ... using priors^$presoftmax_prior_scale_power and rescaling to average 1"
+
```
we have stopped using this in recent scripts.  If you're not using it, please delete the code.

danpovey commented on egs/wsj/s5/steps/nnet3/tdnn/train_raw_nnet.sh:
```diff
@@ -0,0 +1,626 @@
+fi
+
+if $include_log_softmax && [ $stage -le $[$num_iters+1] ]; then
+  echo "Getting average posterior for purposes of adjusting the priors."
```
this whole block is pointless, I think, since we don't need the priors.  Unless you need the average posterior for some other purpose.

vimalmanohar replied:
Yes, I need priors to get likelihoods for use in a HMM decoder for segmentation. I can add a description in the comments.

vimalmanohar commented:
I have made the suggested changes

danpovey commented on egs/wsj/s5/steps/nnet3/tdnn/make_configs.py:
```diff
@@ -357,11 +385,23 @@ def MakeConfigs(config_dir, splice_indexes_string,
     left_context += int(parsed_splice_output['left_context'])
     right_context += int(parsed_splice_output['right_context'])
 
+    add_lda_str = 'false'
```
I'd do add_lda_str = ('true' if add_lda else 'false')

danpovey commented on egs/swbd/s5c/local/nnet3/run_raw_tdnn.sh:
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
```
typically these local/run*.sh script are 'complete', i.e. they contain all the required arguments to actually do something.

vimalmanohar replied:
I understand, but this is meant to be an example for using the configs with the train script. I do not have a complete example yet as the SNR predictor or SAD requires more codes and scripts to be added.

danpovey replied:
maybe explain that it's incomplete and give some idea how you'd use it.

vimalmanohar commented:
I move the example to aspire as an autoencoder mapping from reverberated to clean.

danpovey commented:
OK, merging.

lifeiteng commented:
@vimalmanohar 
From [Softmax Regression vs. k Binary Classifiers] (http://ufldl.stanford.edu/wiki/index.php/Softmax_Regression)
"Now, consider a computer vision example, where you're trying to classify images into three different classes, suppose your classes are indoor_scene, black_and_white_image, and image_has_people. it would be more appropriate to build three separate logistic regression classifiers."

We can make "three separate logistic regression classifiers" deeper, let them share hidden/recurrent layers.

If you have a dataset, each picture is labeled like
* "image has people, not indoor_scene" so you don't kwon wether it is black_and_white_image.

So you can make one positive sample for  image_has_people, one negative sample for indoor_scene, but you can't let it be positive or negative sample for black_and_white_image which means you should't update the separate logistic regression classifier's  parameters of class  black_and_white_image.

vimalmanohar commented:
@lifeiteng 
This can be treated as a standard multi-task problem. I think for this it is most sensible to have two / three separate output nodes in nnet3 configuration and skip specific output nodes from the egs when corresponding label is not defined.
